src/storage/page_buffer.c

```diff
@@ -7345,38 +7372,78 @@ pgbuf_lock_page (THREAD_ENTRY * thread_p, PGBUF_BUFFER_HASH * hash_anchor, const
 		}
-	      pthread_mutex_unlock (&hash_anchor->hash_mutex);
+	      (void) pthread_mutex_unlock (&hash_anchor->hash_mutex);
 	    }
 	  perfmon_inc_stat (thread_p, PSTAT_LK_NUM_WAITED_ON_PAGES);	/* monitoring */
-	  return PGBUF_LOCK_WAITER;
+	  pgbuf_lock_state = PGBUF_LOCK_WAITER;
+	  break;
 	}
       cur_buffer_lock = cur_buffer_lock->lock_next;
     }
 
-  /* buf_lock_table is implemented to have one entry for each thread. At first design, it had one entry for each
-   * thread. cur_thrd_entry->index : thread entry index cur_thrd_entry->tran_index : transaction entry index */
+  if (pgbuf_lock_state != PGBUF_LOCK_WAITER)
+    {
+      /* buf_lock_table is implemented to have one entry for each thread. At first design, it had one entry for each
+       * thread. thread_p->index : thread entry index thread_p->tran_index : transaction entry index */
 
-  /* vpid is not found in the Buffer Lock Chain */
-  cur_buffer_lock = &(pgbuf_Pool.buf_lock_table[cur_thrd_entry->index]);
-  cur_buffer_lock->vpid = *vpid;
-  cur_buffer_lock->next_wait_thrd = NULL;
-  cur_buffer_lock->lock_next = hash_anchor->lock_next;
-  hash_anchor->lock_next = cur_buffer_lock;
-  pthread_mutex_unlock (&hash_anchor->hash_mutex);
+      /* vpid is not found in the Buffer Lock Chain */
+      cur_buffer_lock = &(pgbuf_Pool.buf_lock_table[thread_p->index]);
+      cur_buffer_lock->vpid = *vpid;
+      cur_buffer_lock->next_wait_thrd = NULL;
+      cur_buffer_lock->lock_next = hash_anchor->lock_next;
+      hash_anchor->lock_next = cur_buffer_lock;
+      pthread_mutex_unlock (&hash_anchor->hash_mutex);
+      perfmon_inc_stat (thread_p, PSTAT_LK_NUM_ACQUIRED_ON_PAGES);	/* monitoring */
+      pgbuf_lock_state = PGBUF_LOCK_HOLDER;
+    }
+  assert (pgbuf_lock_state != PGBUF_LOCK_UNKNOWN);
+#else
+  perfmon_inc_stat (thread_p, PSTAT_LK_NUM_ACQUIRED_ON_PAGES);	/* monitoring */
+  pgbuf_lock_state = PGBUF_LOCK_HOLDER;
 #endif /* SERVER_MODE */
 
-  perfmon_inc_stat (thread_p, PSTAT_LK_NUM_ACQUIRED_ON_PAGES);	/* monitoring */
-  return PGBUF_LOCK_HOLDER;
+  if (pgbuf_lock_state == PGBUF_LOCK_WAITER)
+    {
+      if (perf->is_perf_tracking)
+	{
+	  tsc_getticks (&perf->end_tick);
+	  tsc_elapsed_time_usec (&perf->tv_diff, perf->end_tick, perf->start_tick);
+	  perf->lock_wait_time = perf->tv_diff.tv_sec * 1000000LL + perf->tv_diff.tv_usec;
+
+	  if (fetch_mode == NEW_PAGE)
+	    {
+	      perf->perf_page_found = PERF_PAGE_MODE_NEW_LOCK_WAIT;
+	    }
+	  else
+	    {
+	      perf->perf_page_found = PERF_PAGE_MODE_OLD_LOCK_WAIT;
+	    }
+	}
+
+      return pgbuf_lock_state;
+    }
+  assert (pgbuf_lock_state == PGBUF_LOCK_HOLDER);
+
+  if (perf->is_perf_tracking && perf->perf_page_found != PERF_PAGE_MODE_NEW_LOCK_WAIT
+      && perf->perf_page_found != PERF_PAGE_MODE_OLD_LOCK_WAIT)
+    {
+      if (fetch_mode == NEW_PAGE)
+	{
+	  perf->perf_page_found = PERF_PAGE_MODE_NEW_NO_WAIT;
+	}
+      else
+	{
+	  perf->perf_page_found = PERF_PAGE_MODE_OLD_NO_WAIT;
+	}
+    }
```

Comment:
Move code to separate function for readability.
```suggestion
  if (perf->is_perf_tarcking)
    {
      pgbuf_perf_register_page_lock (...);
    }
```

Reply:
Done